### PR TITLE
Fixes #23687 - adds subscription details page

### PR DESCRIPTION
**Pull request by johnpmitsch**
To do:

- [x] ~~add breadcrumb switcher~~ This will actually be handled in http://projects.theforeman.org/issues/22900
- [x] ~~add product content tab (or open issue to work on seperately)~~ will file issue
- [x] linting
- [x] Write tests/create snapshots

### CODE REVIEW

**Comment by waldenraines** [comment]
> add breadcrumb switcher

This will actually be handled in http://projects.theforeman.org/issues/22900

> add product content tab (or open issue to work on seperately)

I'm okay with handling this in a future PR.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        <Table striped bordered condensed hover>
+          <thead>
+            <tr>
+              <td><b>{__("Resource")}</b></td>
+              <td><b>{__("Quantity")}</b></td>
+            </tr>
+          </thead>
+          <tbody>
+            <tr>
+              <td>Content Hosts</td>
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js]: Need translation here

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    super(props);
+  }
+
+  render() {
+    const { subscriptionDetails } = this.props;
+    const searchQuery = 'subscription_id="%s"'.replace('%s', subscriptionDetails.id);
+
+    return (
+      <div>
+        <h2>Associations</h2>
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js]: Need translation here

File: webpack/scenes/Subscriptions/Details/SubscriptionAttributes.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  end_date: "Ends",
+  product_id: "Product ID",
+  contract_number: "Contract Number",
+  account_number: "Account Number",
+  support_level: "Support Level",
+  support_type: "Support Type",
+  arch: "Architecture(s)",
+  type: "Type",
+  mutli_entitlement: "Multi-entitlement",
+  stacking_id: "Stacking ID",
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionAttributes.js]: These should all be translated.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailInfo.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    return __("Yes");
+  }
+
+  render() {
+    const { subscriptionDetails } = this.props;
+
+    return (
+      <div>
+        <h2>Subscription Info</h2>
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailInfo.js]: Need translation here

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailProducts.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    const { subscriptionDetails } = this.props;
+
+    return (
+      <div>
+        <h2>Provided Products</h2>
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailProducts.js]: Translate me please

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailActions.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    })
+    .catch((result) => {
+      dispatch({
+        type: SUBSCRIPTION_DETAILS_FAILURE,
+        error: result.response.data.displayMessage,
+      });
+    });
+}
+ 
+
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailActions.js]: Extra line here

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            <tr>
+              <td>Content Hosts</td>
+              <td>
+                <a href={helpers.urlWithSearch('content_hosts', searchQuery)}>
+                  {subscriptionDetails.host_count}
+                </a>
+              </td>
+            </tr>
+            <tr>
+              <td>Activation Keys</td>
```
Comment (waldenraines) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js]: Translate me

**Comment by waldenraines** [comment]
@rohoover any thoughts on this subscriptions details page?

screenshot from 2018-06-07 10-15-45

The breadcrumbs are missing and will be added in a separate PR.

File: webpack/scenes/Subscriptions/Details/SubscriptionDetails.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import SubscriptionDetailAssociations from './SubscriptionDetailAssociations';
+import SubscriptionDetailProducts from './SubscriptionDetailProducts';
+import { notify } from '../../../move_to_foreman/foreman_toast_notifications';
+
+class SubscriptionDetails extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      loading: true,
```
Comment (amirfefer) [inline webpack/scenes/Subscriptions/Details/SubscriptionDetails.js]: if redux manages this state, you can move it to `initialState`

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailInfo.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    return "";
+  }
+
+  subscriptionDetailValue(subDetails, key) {
+    return subDetails[key] == null ? "" : String(subDetails[key]) 
+  }
+
+  formatInstanceBased(subDetails) {
+    if (subDetails['instance_multiplier'] == null ||
```
[the remark this answers is not in the document]
Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailInfo.js]: This is to check for `null` and `undefined`, which can be checked with the loose equality check of `==`
see [here](https://stackoverflow.com/a/2647888) and [here](https://stackoverflow.com/a/5101991) for more explanation

I can change to dot notation for the properties

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailInfo.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    return __("Yes");
+  }
+
+  render() {
+    const { subscriptionDetails } = this.props;
+
+    return (
+      <div>
+        <h2>Subscription Info</h2>
+        <Table>
```
[the remark this answers is not in the document]
Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailInfo.js]: I'm not sure that component is appropriate here, I am just using the `Table` component for listing the attributes neatly formatted, this isn't a typical Katello/patternfly table listing

File: webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  render() {
+    const { subscriptionDetails } = this.props;
+    const searchQuery = 'subscription_id="%s"'.replace('%s', subscriptionDetails.id);
+
+    return (
+      <div>
+        <h2>Associations</h2>
+        <Table striped bordered condensed hover>
```
[the remark this answers is not in the document]
Comment (johnpmitsch) [reply webpack/scenes/Subscriptions/Details/SubscriptionDetailAssociations.js]: Isn't that too much for what we need here? Its just a simple table with two rows and two columns, not a 'content view'

**Comment by johnpmitsch** [comment]
@amirfefer @waldenraines I updated based on the comments, updates are in a separate commit

**Comment by johnpmitsch** [comment]
> I think this is a big improvement. I am wondering if we should include the data for entitlements out of total. Since that is now reflected in the table.

Since this is a separate API call and the info is available on the table where they can edit the total my vote is to just keep it on the table and not in the details page

**Comment by johnpmitsch** [comment]
Added tests. The only thing missing is the redux reducer/action tests before I can remove WIP. Feel free to review, but please do not review the redux tests yet.

**Comment by johnpmitsch** [comment]
@waldenraines filed the product content issue [here](http://projects.theforeman.org/issues/23886)

**Comment by waldenraines** [comment]
> @amirfefer @waldenraines I updated based on the comments, updates are in a separate commit

Looks good to me, @amirfefer do you mind looking at this once more please?

**Comment by akofink** [comment]
The provided products can get quite long (look at scrollbar in screenshot). Do we want to limit the size of the provided products or maybe overflow with a scroll bar?

image

**Comment by johnpmitsch** [comment]
@Rohoover do you have any suggestions to @akofink's question above? The easiest would be an overflow div with a scrollbar, but that means a scrollbar-inside-of-scrollbar, which I never am a huge fan of. I'm open to any suggestions.

**Comment by akofink** [comment]
Also consider that this is a weird SKU. I don't know if users would *actually* encounter such a long product list.

**Comment by johnpmitsch** [comment]
@akofink that is a good point, I think we can keep the bullet list since our SKUs tend to be considerably longer than users SKUs (afaik). I'm willing to work with @Rohoover if we decide this needs updating later in a follow-up PR.

**Comment by waldenraines** [comment]
[test katello]

**Comment by waldenraines** [comment]
I'm going to go ahead and merge this, mainly so I can close this tab :smile:

**Comment by Rohoover** [comment]
@akofink Two options for you. @waldenraines Normally I'd agree. In this case the list is long and creates a ton of scroll. I think by containing it in a well positioned and spaced pane, we eliminate some of the scroll-in-scroll craziness.

Thoughts?
Sorry for the delay - I was on PTO.

providedproducts
providedproductsv2

**Comment by akofink** [comment]
@Rohoover Thanks for the mockups! I don't think the search is necessary, since we show the entire list (and Ctrl-F should work just fine), but I really like how the pane looks! I'm willing to create a redmine issue for this with the mockup if everyone's happy with it.

**Comment by johnpmitsch** [comment]
+1 to new redmine issue for mockup, thanks @Rohoover!

**Comment by waldenraines** [comment]
> I'm willing to create a redmine issue for this with the mockup if everyone's happy with it.

Works for me, thanks @akofink.  Please link me the issue so I can pull it into kanboard.

**Comment by akofink** [comment]
Redmine: https://projects.theforeman.org/issues/24005

**Pull request merged**